Running tests locally, one test always fails.

I carried out the following to run tests for this project locally:
1. Cloned the repo with submodules so that ably-common was present `git clone git@github.com:ably/ably-go.git --recurse-submodules`
2. Ran the test script `scripts/test.sh --protocol application/json` 
3. Waited for the test run to complete (it took around 14 minutes).

The test results showed a test in `ably/internal/ablyutil` which had failed. The test was called `TestAfterOK` and located in a file called `time_test.go` 

I attempted to rerun this test locally using my IDE
The results of running this test several times were several failures:

```
Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestAfterOK$ github.com/ably/ably-go/ably/internal/ablyutil

=== RUN   TestAfterOK
    /Users/rosiehamilton/go/src/github.com/ably/ably-go/ably/internal/ablyutil/time_test.go:23: expected timer to fire after ~10ms; got 11.007917ms
--- FAIL: TestAfterOK (0.01s)
FAIL
FAIL    github.com/ably/ably-go/ably/internal/ablyutil  0.100s


> Test run finished at 07/02/2022, 16:29:19 <

Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestAfterOK$ github.com/ably/ably-go/ably/internal/ablyutil

=== RUN   TestAfterOK
    /Users/rosiehamilton/go/src/github.com/ably/ably-go/ably/internal/ablyutil/time_test.go:23: expected timer to fire after ~10ms; got 11.019083ms
--- FAIL: TestAfterOK (0.01s)
FAIL
FAIL    github.com/ably/ably-go/ably/internal/ablyutil  0.108s


> Test run finished at 07/02/2022, 16:30:02 <

Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestAfterOK$ github.com/ably/ably-go/ably/internal/ablyutil

=== RUN   TestAfterOK
    /Users/rosiehamilton/go/src/github.com/ably/ably-go/ably/internal/ablyutil/time_test.go:23: expected timer to fire after ~10ms; got 11.014042ms
--- FAIL: TestAfterOK (0.01s)
FAIL
FAIL    github.com/ably/ably-go/ably/internal/ablyutil  0.132s


> Test run finished at 07/02/2022, 16:30:05 <

Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestAfterOK$ github.com/ably/ably-go/ably/internal/ablyutil

=== RUN   TestAfterOK
    /Users/rosiehamilton/go/src/github.com/ably/ably-go/ably/internal/ablyutil/time_test.go:23: expected timer to fire after ~10ms; got 11.010083ms
--- FAIL: TestAfterOK (0.01s)
FAIL
FAIL    github.com/ably/ably-go/ably/internal/ablyutil  0.117s


> Test run finished at 07/02/2022, 16:30:11 <

Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestAfterOK$ github.com/ably/ably-go/ably/internal/ablyutil

=== RUN   TestAfterOK
    /Users/rosiehamilton/go/src/github.com/ably/ably-go/ably/internal/ablyutil/time_test.go:23: expected timer to fire after ~10ms; got 11.011083ms
--- FAIL: TestAfterOK (0.01s)
FAIL
FAIL    github.com/ably/ably-go/ably/internal/ablyutil  0.104s


> Test run finished at 07/02/2022, 16:30:15 <

Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestAfterOK$ github.com/ably/ably-go/ably/internal/ablyutil

=== RUN   TestAfterOK
    /Users/rosiehamilton/go/src/github.com/ably/ably-go/ably/internal/ablyutil/time_test.go:23: expected timer to fire after ~10ms; got 11.026875ms
--- FAIL: TestAfterOK (0.01s)
FAIL
FAIL    github.com/ably/ably-go/ably/internal/ablyutil  0.124s


> Test run finished at 07/02/2022, 16:32:28 <

Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestAfterOK$ github.com/ably/ably-go/ably/internal/ablyutil

=== RUN   TestAfterOK
    /Users/rosiehamilton/go/src/github.com/ably/ably-go/ably/internal/ablyutil/time_test.go:23: expected timer to fire after ~10ms; got 11.013209ms
--- FAIL: TestAfterOK (0.01s)
FAIL
FAIL    github.com/ably/ably-go/ably/internal/ablyutil  0.216s


> Test run finished at 07/02/2022, 16:34:54 <
```

It appears this test is testing that a timer fires within 1ms of an expected time, but when running locally this always seems to takes around 1.01ms.

This test will be passing inside the CI pipeline as I used the same command to run them locally that CI uses.

However, a developer should be able to run tests locally and get the same result as the test run inside CI. This test should be fixed so that it works locally.

This will start to give us certainty when running tests locally and help us build and place trust in these tests.

┆sprint: Backlog

--- a/ably/internal/ablyutil/time_test.go
+++ b/ably/internal/ablyutil/time_test.go
@@ -59,6 +59,6 @@ func TestAfterCanceled(t *testing.T) {
 }
 
 func isCloseTo(d time.Duration, to time.Duration) bool {
-	const leeway = 1 * time.Millisecond
+	const leeway = 2 * time.Millisecond
 	return d > to-leeway && d < to+leeway
 }

